Code/Helpers/mpi_pccg_solvers.c

```diff
@@ -0,0 +1,116 @@
+#ifndef MPI_PCCG_SOLVERS_C
+#define MPI_PCCG_SOLVERS_C
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <math.h>
+#include <mpi.h>
+#include <string.h>
+
+#include "poisson_matrix_creaters.c"
+#include "matrix_arithmetic.c"
+#include "helpers.c"
+#include "preconditioner_solve.c"
+#include "mpi_helpers.c"
+
+struct CGret mpiPCCG_solveCSR(struct par_multdat pmd, struct A_csr A, float* b, float* xg, float tol, char* pctype)
+{
+    int rank = pmd.rank_d;
+    int n = pmd.n_d;
+    struct CGret cgret;
+    struct PCret pcret;
+    float rsold, rsnew, alpha, rnorm;
+    float* Ap = create1dZeroVec(n);
+    float* r = create1dZeroVec(n); 
+    float* dum = create1dZeroVec(n);
+    float* p = create1dZeroVec(n); 
+    float* x = create1dZeroVec(n); 
+    float* z = create1dZeroVec(n);
+
+    x = VecAdd1(n,xg,r,0);
+    dum = mpiMatVecProductCSR1(pmd,xg,A);
+    r = VecAdd1(n,b,dum,-1); //r=b-A*x
+    pcret = PC_Solve(n,r,A,pctype); //Solve M*z0=r0
+    z = pcret.sol;
+    p = z;
+
+    rsold = innerProd1(n,r,z); //rsold=r'*r
+    int iters = 0;
+    for (int i=0; i<n; i++){
+        Ap = mpiMatVecProductCSR1(pmd,p,A); // Ap=A*p
+        alpha = rsold / innerProd1(n,p,Ap); // alpha=rsold/(p'*Ap)
+        x = VecAdd1(n,x,p,alpha); // x=x+alpha*p
+        r = VecAdd1(n,r,Ap,-alpha); // r=r-alpha*Ap
+        rnorm = innerProd1(n,r,r); // rnorm=r'*r;
+        if (sqrt(rnorm) < tol) break; // Check convergence
+        pcret = PC_Solve(n,r,A,pctype); //Solve M*z=r
```

Comment:
This looks to me like we can add to the `PC_Solve` function our other preconditioners. Nice work!